File: app/csharp/src/Utils/transfer_client.cs
using Grpc.Net.Client;
using Newtonsoft.Json.Linq;
using System;
using System.IO;
using System.Text.RegularExpressions;
/// <summary>
/// Provides the following services:
/// <list type="bullet">
/// <item>
/// <description>daemon conf file generation, startup and shutdown of transferd</description>
/// </item>
/// <item>
/// <description>transfer of files and monitoring</description>
/// </item>
/// </list>
/// </summary>
public class TransferClient
{
    private const string ASCP_LOG_FILE = "aspera-scp-transfer.log";
    private Configuration _config;
    private string _serverAddress;
    private int _serverPort;
    private System.Diagnostics.Process _daemonProcess = null;
    private List<StreamWriter> _daemonStreams = new List<StreamWriter>();
    private Transferd.Api.TransferService.TransferServiceClient _daemonService = null;
    private string _daemonName;
    private string _daemonLog;


    public TransferClient(Configuration config)
    {
        _config = config;
        var confUrl = new Uri(_config.GetParam("trsdk", "url"));
        _serverAddress = confUrl.Host;
        _serverPort = confUrl.Port;
        _daemonName = Path.GetFileName(_config.GetPath("sdk_daemon"));
        _daemonLog = Path.Combine(_config.LogFolder(), _daemonName + ".log");
    }

    public void CreateConfigFile(string confFile)
    {
        var configInfo = new
        {
            address = _serverAddress,
            port = _serverPort,
            log_directory = _config.LogFolder(),
            log_level = _config.GetParam("trsdk", "level"),
            fasp_runtime = new
            {
                use_embedded = true,
                log = new
                {
                    dir = _config.LogFolder(),
                    level = AscpLevel(_config.GetParam("trsdk", "ascp_level")),
                },
            },
        };
        File.WriteAllText(confFile, Newtonsoft.Json.JsonConvert.SerializeObject(configInfo));
    }

    /// <summary>
    /// Start transfer manager daemon if not already running and return gRPC client
    /// </summary>
    /// <exception cref="Exception"></exception>
    public void StartDaemon()
    {
        Log.log.Info("ERROR: Failed to connect\nStarting daemon...");
        var daemonPath = _config.GetPath("sdk_daemon");
        var fileBase = Path.Combine(_config.LogFolder(), _daemonName);
        var confFile = fileBase + ".conf";
        var outFile = fileBase + ".out";
        var errFile = fileBase + ".err";
        var exec_args = $"--config {confFile}";
        var command = $"{daemonPath} {exec_args}";
        Log.log.Debug($"daemon out: {outFile}");
        Log.log.Debug($"daemon err: {errFile}");
        Log.log.Debug($"daemon log: {_daemonLog}");
        Log.log.Debug($"ascp log: {Path.Combine(_config.LogFolder(), ASCP_LOG_FILE)}");
        Log.log.Debug($"command: {command}");
        CreateConfigFile(confFile);
        Log.log.Info("Starting daemon...");
        _daemonProcess = new System.Diagnostics.Process
        {
            StartInfo = new System.Diagnostics.ProcessStartInfo
            {
                FileName = daemonPath,
                Arguments = exec_args,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            }
        };
        _daemonProcess.OutputDataReceived += captureStream(outFile);
        _daemonProcess.ErrorDataReceived += captureStream(errFile);
        _daemonProcess.Start();
        // wait for daemon to be ready
        Thread.Sleep(2000);
        if (_daemonProcess.HasExited)
        {
            Log.log.Error($"Daemon not started.");
            Log.log.Error($"Exited with code: {_daemonProcess.ExitCode}");
            Log.log.Error($"Check daemon log: {_daemonLog}");
            _daemonProcess.WaitForExit();
            _daemonProcess = null;
            //logging.error(utils.configuration.last_file_line(self._daemon_log));
            throw new Exception("daemon startup failed");
        }
        _daemonProcess.BeginOutputReadLine();
        _daemonProcess.BeginErrorReadLine();

        if (_serverPort == 0)
        {
            string lastLine = File.ReadLines(_daemonLog).Last();
            JObject logInfo = JObject.Parse(lastLine);
            Match portMatch = Regex.Match(logInfo["msg"]?.ToString() ?? "", @":(\d+)");

            if (!portMatch.Success)
            {
                throw new Exception("Could not read listening port from log file");
            }

            _serverPort = int.Parse(portMatch.Groups[1].Value);
            Log.log.Debug($"Allocated server port: {_serverPort}");
        }
    }
    public void ConnectToDaemon()
    {
        AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
        var grpcUrl = new Uri($"http://{_serverAddress}:{_serverPort}");
        Log.log.Info($"Connecting to {_daemonName} on {grpcUrl} ...");
        _daemonService = new Transferd.Api.TransferService.TransferServiceClient(GrpcChannel.ForAddress(grpcUrl));
        _daemonService.GetAPIVersion(new Transferd.Api.APIVersionRequest());
        Log.log.Info("Connected !");
    }
    public void Startup()
    {
        if (_daemonService == null)
        {
            StartDaemon();
            ConnectToDaemon();
        }
    }
    /// <summary>
    /// Shutdown transfer manager daemon, if needed
    /// </summary>
    public void Shutdown()
    {
        _daemonService = null;
        // Shutdown transfer manager daemon, if needed
        if (_daemonProcess != null)
        {
            Log.log.Info("Stopping Transfer daemon...");
            _daemonProcess.Kill();
            _daemonProcess.WaitForExit();
            _daemonProcess = null;
            Log.log.Info("Transfer daemon has been terminated.");
            foreach (var stream in _daemonStreams)
            {
                stream.Close();
            }
        }
    }

    /// <summary>
    /// Start the specified transfer
    /// </summary>
    /// <param name="aSpecObj">transfer specification (JSON Object)</param>
    /// <returns>transfer id</returns>
    public string StartTransfer(JObject aSpecObj)
    {
        Log.log.Info(aSpecObj);
        // Start a transfer and return transfer id
        var transferRequest = new Transferd.Api.TransferRequest
        {
            TransferType = Transferd.Api.TransferType.FileRegular,
            Config = new Transferd.Api.TransferConfig { LogLevel = 2 },
            TransferSpec = Newtonsoft.Json.JsonConvert.SerializeObject(aSpecObj),
        };

        var transferResponse = _daemonService.StartTransfer(transferRequest);

        if (transferResponse.Status == Transferd.Api.TransferStatus.Failed)
        {
            Log.log.Info($"ERROR: {transferResponse.Error.Description}");
            Environment.Exit(1);
        }

        return transferResponse.TransferId;
    }

    /// <summary>
    /// wait until the specified transfer is finished (completed or failed)
    /// </summary>
    /// <param name="aTransferId"></param>
    void WaitTransfer(string aTransferId)
    {
        while (true)
        {
            // check the current state of the transfer
            var queryTransferResponse = _daemonService.QueryTransfer(new Transferd.Api.TransferInfoRequest() { TransferId = aTransferId });
            Console.Out.WriteLine("transfer info " + queryTransferResponse);

            // check transfer status in response, and exit if it's done
            Transferd.Api.TransferStatus status = queryTransferResponse.Status;
            if (status == Transferd.Api.TransferStatus.Failed || status == Transferd.Api.TransferStatus.Completed)
            {
                Console.Out.WriteLine("finished " + status);
                break;
            }
            // wait a second before checking again
            System.Threading.Thread.Sleep(1000);
        }
    }


    /// <summary>
    /// One-call simplified procedure to start daemon, transfer, and wait for it to finish
    /// </summary>
    /// <param name="aSpecObj">transfer specification (JSON Object)</param>
    public void StartTransferAndWait(JObject aSpecObj)
    {
        Startup();
        WaitTransfer(StartTransfer(aSpecObj));
    }
    /// <summary>
    /// Capture stdout or stderr for the started process (transferd)
    /// </summary>
    /// <param name="logFile"></param>
    /// <returns></returns>
    public System.Diagnostics.DataReceivedEventHandler captureStream(string logFile)
    {
        var logStream = new StreamWriter(new FileStream(logFile, FileMode.Append, FileAccess.Write));
        _daemonStreams.Add(logStream);
        return new System.Diagnostics.DataReceivedEventHandler(
            (sender, e) =>
            {
                if (!String.IsNullOrEmpty(e.Data))
                {
                    logStream.WriteLine(e.Data);
                }
            });
    }
    /// <summary>
    /// translates string log level to numerical
    /// </summary>
    /// <param name="level">text level</param>
    /// <returns>numerical value</returns>
    /// <exception cref="ArgumentException"></exception>
    private static int AscpLevel(string level)
    {
        if (level == "info")
        {
            return 0;
        }
        else if (level == "debug")
        {
            return 1;
        }
        else if (level == "trace")
        {
            return 2;
        }
        else
        {
            throw new ArgumentException("Invalid ascp_level: " + level);
        }
    }
}

File: app/go/src/utils/transfer_client.go
// Simplified access to Aspera Transfer Daemon client
package utils

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"time"

	pb "aspera_examples/build/grpc_aspera"

	"go.uber.org/zap"
	"google.golang.org/grpc"
)

const (
	ASCP_LOG_FILE = "aspera-scp-transfer.log"
)

type TransferClient struct {
	config             *Configuration
	serverAddress      string
	serverPort         int
	transferDaemonProc *exec.Cmd
	transferService    pb.TransferServiceClient
	daemonName         string
	daemonLog          string
}

func NewTransferClient(config *Configuration) *TransferClient {
	sdkURL, err := url.Parse(config.ParamStr("trsdk", "url"))
	if err != nil {
		config.Log.Fatalf("Error parsing server URL: %v", err)
	}
	return &TransferClient{
		config:        config,
		daemonName:    filepath.Base(config.GetPath("sdk_daemon")),
		daemonLog:     filepath.Join(config.LogFolder, filepath.Base(config.GetPath("sdk_daemon"))+".log"),
		serverAddress: sdkURL.Hostname(),
		serverPort:    GetPortOrDefault(sdkURL, 33001),
	}
}

func (tc *TransferClient) CreateConfigFile(confFile string) error {
	ascpLevel := tc.config.ParamStr("trsdk", "ascp_level")
	var ascpIntLevel int
	switch ascpLevel {
	case "info":
		ascpIntLevel = 0
	case "debug":
		ascpIntLevel = 1
	case "trace":
		ascpIntLevel = 2
	default:
		return fmt.Errorf("invalid ascp_level: %s", ascpLevel)
	}

	configInfo := map[string]interface{}{
		"address":       tc.serverAddress,
		"port":          tc.serverPort,
		"log_directory": tc.config.LogFolder,
		"log_level":     tc.config.ParamStr("trsdk", "level"),
		"fasp_runtime": map[string]interface{}{
			"use_embedded": true,
			"log": map[string]interface{}{
				"dir":   tc.config.LogFolder,
				"level": ascpIntLevel,
			},
		},
	}

	configData, err := json.Marshal(configInfo)
	if err != nil {
		return err
	}

	tc.config.Log.Debugf("config: %s", string(configData))
	return os.WriteFile(confFile, configData, 0644)
}

func (tc *TransferClient) StartDaemon() error {
	confFile := filepath.Join(tc.config.LogFolder, tc.daemonName+".conf")
	outFile := filepath.Join(tc.config.LogFolder, tc.daemonName+".out")
	errFile := filepath.Join(tc.config.LogFolder, tc.daemonName+".err")
	cmd := exec.Command(
		tc.config.GetPath("sdk_daemon"),
		"--config", confFile,
	)

	tc.CreateConfigFile(confFile)

	tc.config.Log.Info("Starting daemon...", zap.String("command", cmd.String()))

	// Redirection des logs
	cmd.Stdout = tc.openFile(outFile)
	cmd.Stderr = tc.openFile(errFile)

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("failed to start daemon: %w", err)
	}

	tc.transferDaemonProc = cmd
	time.Sleep(2 * time.Second)

	if err := tc.checkDaemonStartup(); err != nil {
		return err
	}

	return nil
}

func (tc *TransferClient) checkDaemonStartup() error {
	if tc.transferDaemonProc.ProcessState != nil && tc.transferDaemonProc.ProcessState.Exited() {
		tc.config.Log.Error("Daemon not started", zap.Error(fmt.Errorf("exit status: %v", tc.transferDaemonProc.ProcessState.ExitCode())))
		return fmt.Errorf("daemon startup failed")
	}

	if tc.serverPort == 0 {
		logLine, err := LastFileLine(tc.daemonLog)
		if err != nil {
			return err
		}
		tc.config.Log.Debugf("Last log line: %s", logLine)
		re := regexp.MustCompile(`:(\d+)`)
		match := re.FindStringSubmatch(logLine)
		if match == nil {
			return fmt.Errorf("could not read listening port from log file")
		}
		port, err := strconv.Atoi(match[1])
		if err != nil {
			return fmt.Errorf("could not parse port number: %w", err)
		}

		tc.serverPort = port
		tc.config.Log.Infof("Allocated server port : %d", tc.serverPort)
	}

	return nil
}

func (tc *TransferClient) ConnectToDaemon() error {
	address := fmt.Sprintf("%s:%d", tc.serverAddress, tc.serverPort)
	tc.config.Log.Info("Connecting to transfer daemon...", zap.String("address", address))

	channel, err := grpc.NewClient(address, grpc.WithInsecure())
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}

	tc.transferService = pb.NewTransferServiceClient(channel)
	tc.config.Log.Info("Connected!")
	return nil
}

func (tc *TransferClient) Startup() error {
	if tc.transferService == nil {
		if err := tc.StartDaemon(); err != nil {
			return err
		}
		if err := tc.ConnectToDaemon(); err != nil {
			return err
		}
	}
	return nil
}

func (tc *TransferClient) Shutdown() error {
	if tc.transferDaemonProc != nil {
		tc.config.Log.Info("Shutting down daemon...")
		return tc.transferDaemonProc.Process.Kill()
	}
	return nil
}

func (tc *TransferClient) StartTransfer(transferSpec map[string]interface{}) (string, error) {
	tsJSON, err := json.Marshal(transferSpec)
	if err != nil {
		return "", fmt.Errorf("failed to marshal transfer spec: %w", err)
	}

	tc.config.Log.Debugf("Transfer spec: %s", string(tsJSON))

	req := &pb.TransferRequest{
		TransferType: pb.TransferType_FILE_REGULAR,
		Config:       &pb.TransferConfig{},
		TransferSpec: string(tsJSON),
	}

	resp, err := tc.transferService.StartTransfer(context.TODO(), req)
	if err != nil {
		return "", fmt.Errorf("failed to start transfer: %w", err)
	}

	if err := tc.throwOnError(resp.Status, resp.Error); err != nil {
		return "", err
	}

	return resp.TransferId, nil
}

func (tc *TransferClient) WaitTransfer(transferID string) error {
	req := &pb.RegistrationRequest{
		Filters: []*pb.RegistrationFilter{
			{TransferId: []string{transferID}},
		},
	}

	stream, err := tc.transferService.MonitorTransfers(context.Background(), req)
	if err != nil {
		return fmt.Errorf("failed to monitor transfer: %w", err)
	}

	for {
		info, err := stream.Recv()
		if err != nil {
			return fmt.Errorf("failed to receive transfer info: %w", err)
		}

		tc.config.Log.Info("Transfer status", zap.String("status", pb.TransferStatus_name[int32(info.Status)]))

		if err := tc.throwOnError(info.Status, info.Error); err != nil {
			return err
		}

		if info.Status == pb.TransferStatus_COMPLETED {
			break
		}
	}

	return nil
}

func (tc *TransferClient) StartTransferAndWait(transferSpec map[string]interface{}) error {
	if err := tc.Startup(); err != nil {
		return err
	}

	transferID, err := tc.StartTransfer(transferSpec)
	if err != nil {
		return err
	}

	return tc.WaitTransfer(transferID)
}

func (tc *TransferClient) throwOnError(status pb.TransferStatus, error *pb.Error) error {
	switch status {
	case pb.TransferStatus_FAILED:
		tc.config.Log.Errorf("Transfer failed: %s", error.Description)
		return fmt.Errorf("transfer failed: %s", error.Description)
	case pb.TransferStatus_UNKNOWN_STATUS:
		return fmt.Errorf("unknown transfer status: %s", error.Description)
	default:
		return nil
	}
}

// Utilitaire pour ouvrir les fichiers de log
func (tc *TransferClient) openFile(filename string) *os.File {
	file, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		tc.config.Log.Fatalf("Failed to open log file: %s: %s", filename, err)
	}
	return file
}
